lib/server.ml

```diff
@@ -36,70 +36,115 @@ let make channel =
+    chan.read req
     >>= fun () ->
     match OptionRequestHeader.unmarshal req with
     | Error e -> fail e
     | Ok hdr ->
-      let payload = Cstruct.create (Int32.to_int hdr.OptionRequestHeader.length) in
-      channel.read payload
+      let payload = make_blank_payload hdr in
+      chan.read payload
       >>= fun () ->
       begin match hdr.OptionRequestHeader.ty with
-        | Option.ExportName -> return (Cstruct.to_string payload, make channel)
+        | Option.StartTLS as opt ->
+          let resp = if chan.is_tls then OptionResponse.Invalid else OptionResponse.Policy in
+          respond opt resp chan.write
+          >>= fun () ->
+          loop ()
+        | Option.ExportName -> return (Cstruct.to_string payload, make chan)
         | Option.Abort -> fail (Failure "client requested abort")
-        | Option.Unknown x ->
-          OptionResponseHeader.(marshal res { request_type = Option.Unknown x; response_type = OptionResponse.Unsupported; length = 0l });
-          channel.write res
+        | Option.Unknown _ as opt ->
+          respond opt OptionResponse.Unsupported chan.write
           >>= fun () ->
           loop ()
-        | Option.List ->
+        | Option.List as opt ->
           begin match offer with
             | None ->
-              OptionResponseHeader.(marshal res { request_type = Option.List; response_type = OptionResponse.Policy; length = 0l });
-              channel.write res
+              respond opt OptionResponse.Policy chan.write
               >>= fun () ->
               loop ()
             | Some offers ->
               let rec advertise = function
-                | [] ->
-                  OptionResponseHeader.(marshal res { request_type = Option.List; response_type = OptionResponse.Ack; length = 0l });
-                  channel.write res
+                | [] -> send_ack opt chan.write
                 | x :: xs ->
                   let l = String.length x in
-                  OptionResponseHeader.(marshal res { request_type = Option.List; response_type = OptionResponse.Server; length = Int32.of_int l});
-                  channel.write res
+                  OptionResponseHeader.(marshal res { request_type = opt; response_type = OptionResponse.Server; length = Int32.of_int l});
+                  chan.write res
                   >>= fun () ->
                   let name = Cstruct.create l in
                   Cstruct.blit_from_string x 0 name 0 l;
-                  channel.write name
+                  chan.write name
                   >>= fun () ->
                   advertise xs in
               advertise offers
               >>= fun () ->
               loop ()
           end
-      end in
-  loop ()
+      end
+    in loop ()
+  in
+  let rec negotiate_tls () =
+    channel.read_clear req
+    >>= fun () ->
+    match OptionRequestHeader.unmarshal req with
+    | Error e -> fail e
+    | Ok hdr ->
+      let payload = make_blank_payload hdr in
```

Comment:
In the match cases below, the payload length should be 0 for `NBD_OPT_STARTTLS`, and there should be no payload. We could add some more checks here to cause buggy clients to fail earlier, but it's outside the scope of this PR and not important to add here.